tests/sqlx/lexer.spec.ts

```diff
@@ -3,20 +3,42 @@ import { expect } from "chai";
 
 describe("@dataform/sqlx", () => {
   describe("lexer", () => {
-    it("replaces_single_backslash_with_double_backslashes", () => {
+    it("replaces single backslash with double backslashes", () => {
       // These look like groups of 2 backslashes, but are actually 1. If someone
       // in SQLX SQL writes "\" in the regex, it should be interpreted as "\".
       const parsedSqlx = parseSqlx("SELECT regexp_extract('01a_data_engine', '^(\\d{2}\\w)')");
-      expect(parsedSqlx.sql[0]).equals(
-        "SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')"
-      );
+      expect(parsedSqlx.sql).eql(["SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')"]);
     });
-    it("replaces_double_backslashes_with_quadruple_backslashes", () => {
+    it("replaces double backslashes with quadruple backslashes", () => {
       // In SQLX SQL, "\\" should be interpreted as "\\".
       const parsedSqlx = parseSqlx("SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')");
-      expect(parsedSqlx.sql[0]).equals(
+      expect(parsedSqlx.sql).eql([
         "SELECT regexp_extract('01a_data_engine', '^(\\\\\\\\d{2}\\\\\\\\w)')"
+      ]);
+    });
+    it("replaces single backslash with double backslashes in pre operation", () => {
+      const parsedSqlx = parseSqlx(
+        "pre_operations {SELECT regexp_extract('01a_data_engine', '^(\\d{2}\\w)')}"
       );
+      expect(parsedSqlx.preOperations).eql([
+        "SELECT regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')"
+      ]);
+    });
+    it("closing braces in inner sql quotes doesn't close block", () => {
+      const parsedSqlx = parseSqlx("pre_operations {SELECT '{}'}");
+      expect(parsedSqlx.preOperations).eql(["SELECT '{}'"]);
+    });
+    it("closing braces in inner sql quotes with backslashes after doesn't close block", () => {
+      const parsedSqlx = parseSqlx("pre_operations {SELECT '\\d{2}'}");
+      expect(parsedSqlx.preOperations).eql(["SELECT '\\\\d{2}'"]);
+    });
+    it("escaped quotes do not close single quoted string", () => {
```

Comment:
i'm not super sure what this test case is asserting. how would we know if it's closing the string or not? i think you might want to add some "special" text inside a still-open string, e.g. `js { foo = "bar" }` and see that the whole text is returned.

in general, since the PR changes how we lex strings fairly significantly, i think it would be a good idea to add some test cases for string matching - asserting that escaped end-quotes don't end the string, but also handling escaped backslashes, then escaped backslashes next to an un-escaped end-quote, etc.

Reply:
You are right, special text breaks it

Reply:
<img width="480" alt="Screenshot 2019-12-18 at 16 39 31" src="https://user-images.githubusercontent.com/13434377/71105182-00890b80-21b5-11ea-8268-3af977885f58.png">

Reply:
Issue was I hadn't added string handling to outer SQL, just inner SQL blocks. Have added more thorough tests now to make sure.